packages/plugin-electron-device/device.js

```diff
@@ -45,15 +45,51 @@ module.exports = (app, screen, process, filestore, NativeClient, powerMonitor) =
       }
     })
 
-    client.addMetadata('device', {
-      // the value for 'idleThreshold' doesn't matter here because it is ignored
-      // if the system is locked and that's the only state we care about
-      isLocked: powerMonitor.getSystemIdleState(1) === 'locked',
-      // note: this is only available from Electron 12, earlier versions cannot
-      // read the battery state without 'on-ac'/'on-battery' events
-      usingBattery: powerMonitor.onBatteryPower
+    app.whenReady().then(() => {
+      client.addMetadata('device', {
+        // the value for 'idleThreshold' doesn't matter here because it is ignored
+        // if the system is locked and that's the only state we care about
+        isLocked: powerMonitor.getSystemIdleState(1) === 'locked',
+        // note: this is only available from Electron 12, earlier versions cannot
+        // read the battery state without 'on-ac'/'on-battery' events
+        usingBattery: powerMonitor.onBatteryPower
+      })
+
+      powerMonitor.on('on-ac', () => {
+        client.addMetadata('device', { usingBattery: false })
+      })
+
+      powerMonitor.on('on-battery', () => {
+        client.addMetadata('device', { usingBattery: true })
+      })
+
+      powerMonitor.on('unlock-screen', () => {
+        client.addMetadata('device', { isLocked: false })
+      })
+
+      powerMonitor.on('lock-screen', () => {
+        client.addMetadata('device', { isLocked: true })
+      })
     })
 
+    client.addOnError(event => {
+      event.device = Object.assign(
+        {},
+        event.device,
+        device,
+        {
+          freeMemory: kibibytesToBytes(process.getSystemMemoryInfo().free),
+          time: new Date()
+        }
+      )
+
+      if (powerMonitor && powerMonitor.getSystemIdleTime) {
+        event.addMetadata('device', { idleTime: powerMonitor.getSystemIdleTime() })
+      }
+
+      setDefaultUserId(event)
```

Comment:
I don't think this should be inside the `whenReady` as it'll mean events that are triggered before the app is ready will be missing a user ID

Reply:
This isn't part of the `whenReady`, that ends on line 73 and the next one picks up on line 93.